ILMarshaler Refactor

jkoritzinsky opened the pull request:
Refactor the interop `ILMarshaler`s to have a more consistent structure. Add some comments for how the marshaler system is structured.

This could be a first step towards unifying our parameter/return value marshalling with our field marshalling. Or at least towards making the code easier to reason about for any future path we take with the marshalers.

jkoritzinsky commented:
@dotnet-bot help

jkoritzinsky commented:
@dotnet-bot test this please

AaronRobinsonMSFT commented on src/vm/ilmarshalers.h:
```diff
@@ -1417,7 +1474,16 @@ class ILMarshaler
 public:
+    
+    // Extension point to allow a marshaler to conditionally override all of the ILMarshaler logic with its own or block marshalling when marshalling an argument.
+    // See mlinfo.cpp:3523 for the implementation.
```
Like the comment - remove the line number and replace with function name.

AaronRobinsonMSFT replied:
I think there are a few other spots this happened - i.e. line number.

jkoritzinsky replied:
I only remember doing that on this line and again a few lines lower. I'll go verify now.

jkoritzinsky replied:
Yeah those were the only two cases of that in this PR.

AaronRobinsonMSFT commented on src/vm/ilmarshalers.h:
```diff
@@ -1781,73 +1847,73 @@ class ILReflectionObjectMarshaler : public ILMarshaler
 
     virtual CorElementType GetNativeBoolElementType() = 0;
-    virtual int GetNativeTrueValue() = 0;
+    virtual int GetNativeTrueValue() = 0 ;
```
Rogue space?

jkoritzinsky commented:
@dotnet-bot test this please

jkoritzinsky commented:
Removed some changes from this PR that aren't as useful and rebased on master. 

Diff is significantly easier to read with whitespace changes ignored (approximately half the size of what it was before the rebase).

AaronRobinsonMSFT commented on src/vm/ilmarshalers.h:
```diff
@@ -2832,26 +2939,42 @@ class ILAsAnyMarshalerBase : public ILMarshaler
+    }
+
+private:
+    DWORD GetAsAnyFlags()
```
can be marked `const`?

AaronRobinsonMSFT commented on src/vm/ilmarshalers.h:
```diff
@@ -2832,26 +2939,42 @@ class ILAsAnyMarshalerBase : public ILMarshaler
-    }
-
 protected:
     static const BYTE ML_IN  = 0x10;
```
Can you move both `ML_IN` and `ML_OUT` to usage and document why the values are what they are.

jkoritzinsky commented:
The CI failures are a known issue. I'll retrigger CI once they're fixed and run some tests on some WinForms and WPF apps to validate behavior.

jkoritzinsky commented:
I've run this branch against the WinForms integration test suite and against one apps in the WPF AppCompat suite and everything works as expected. If CI is green after I address your comments, I'll merge this in.

AaronRobinsonMSFT commented:
@stephentoub This PR inserted the behavioral change you are now observing regarding excessive allocations.

@jkoritzinsky This has severely impacted a scenario involving `HttpClient`. We need to either fix the regression or back this out ASAP. Based on @stephentoub's allocation profile the underlying issue appears to be excessive allocations in the code below:

https://github.com/dotnet/coreclr/blob/f163abbd16d35032e4450a09c548cb804cf36a85/src/System.Private.CoreLib/src/System/StubHelpers.cs#L1558-L1563

stephentoub commented:
I'm seeing these getting allocated now for every input SafeHandle marshaled, e.g. for every call we make on a socket.  Is that expected?  We need to not do that ;)
cc: @jkotas
